src/core/meshcop/joiner.cpp

```diff
@@ -86,6 +86,30 @@ void Joiner::SetState(otJoinerState aState)
     return;
 }
 
+otError Joiner::ValidatePskd(const char *aPskd)
+{
+    otError error      = OT_ERROR_INVALID_ARGS;
+    size_t  pskdLength = strlen(aPskd);
+
+    OT_STATIC_ASSERT(static_cast<uint8_t>(kPskdMaxLength) <= static_cast<uint8_t>(Dtls::kPskMaxLength),
+                     "The maximum length of DTLS PSK is smaller than joiner PSKd");
+
+    VerifyOrExit(pskdLength >= kPskdMinLength && pskdLength <= kPskdMaxLength, OT_NOOP);
+
+    for (size_t i = 0; i < pskdLength; i++)
+    {
+        char c = aPskd[i];
+
+        VerifyOrExit(isdigit(c) || isupper(c), OT_NOOP);
+        VerifyOrExit(c != 'I' && c != 'O' && c != 'Q' && c != 'Z', OT_NOOP);
```

Comment:
What about using `switch` here to make it a little bit faster?

Reply:
I don't think performance matters here, since the joiner API is not supposed to be called frequently.